src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py

```diff
@@ -64,75 +73,108 @@ def save_pretrained_wrapper(
+                        tokenizer=tokenizer,
+                    )
+                    # only allow the main process move the state
+                    # dicts to cpu
+                    if trainer.accelerator.is_main_process:
+                        # assuming quantization is the last step
+                        # we no longer need the original model
+                        # and can safely delete it to save memory
+                        del trainer.model
+                        find_and_move_state_dicts_to_cpu(save_directory)
+
+            else:
+                # HACK: Override the dtype_byte_size function in transformers to
+                # support float8 types. Fix is posted upstream
+                # https://github.com/huggingface/transformers/pull/30488
+                transformers.modeling_utils.dtype_byte_size = new_dtype_byte_size
+
+                model = trainer.model
+                # state_dict gets passed in as a kwarg for FSDP models
+                state_dict = kwargs.pop("state_dict", None)
+
+                # find offloaded state dict if none is provided
+                if state_dict is None:
+                    state_dict = get_state_dict_offloaded_model(model)
+
+                if sparsity_config is not None:
+                    sparsity_config.global_sparsity = (
+                        SparsityConfigMetadata.infer_global_sparsity(
+                            model, state_dict=state_dict
+                        )
+                    )
+                    sparsity_config.sparsity_structure = (
+                        SparsityConfigMetadata.infer_sparsity_structure()
+                    )
+                elif not skip_compression_stats:
+                    # try to infer a sparsity config from the model if none is provided
+                    logger.info(
+                        "Inferring a sparsity configuration requires a global sparsity "
+                        "calculation. This can be costly for large models. To skip the "
+                        "calculation of compression statistics set "
+                        "skip_compression_stats=True"
+                    )
+                    sparsity_config = SparsityConfigMetadata.from_pretrained(
+                        model, state_dict=state_dict, compress=save_compressed
+                    )
 
-            # HACK: Override the dtype_byte_size function in transformers to
-            # support float8 types. Fix is posted upstream
-            # https://github.com/huggingface/transformers/pull/30488
-            transformers.modeling_utils.dtype_byte_size = new_dtype_byte_size
+                quantization_format = infer_quantization_format(
+                    model=model,
+                    quantization_format=quantization_format,
+                    save_compressed=save_compressed,
+                    sparsity_config=sparsity_config,
+                )
+                compressor = ModelCompressor.from_pretrained_model(
```

Comment:
We only want this method to be responsible for saving the files to disk post compression. The ModelCompressor lifecyle should be outside of save_pretrained.

Reply:
Next to dos, updated below